src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -14506,6 +15404,255 @@ public static Vector512<T> Invoke(Vector512<T> x, Vector512<T> y)
             public static Vector512<T> Invoke(Vector512<T> x) => (x * T.CreateChecked(180)) / T.Pi;
         }
 
+        /// <summary>T &lt;&lt; amount</summary>
+        internal readonly struct ShiftLeftOperator<T>(int amount) : IStatefulUnaryOperator<T> where T : IBinaryInteger<T>
+        {
+            private readonly int _amount = amount;
```

Comment:
Also, is the JIT correctly inlining this and hoisting `_amount` out of the loop so that it doesn't broadcast or create scalar before every shift operation?

That is, most hardware has two-three forms of the shift instructions. For x64, we have:
```csharp
vpslld   xmm1, xmm2,  imm8       ; `xmm1[i] = xmm2[i] << imm8`
vpslld   xmm1, xmm2, xmm3/m128   ; `xmm1[i] = xmm2[i] << xmm2.AsUInt64().ToScalar()`
vpsllvd xmm1, xmm2, xmm3/m128    ; `xmm1[i] = xmm2[i] << xmm3[i]`
```

So we have to emit one of the latter two encodings for this case and should be picking the second form, afair. But that requires we internally do the equivalent of `Sse2.ShiftLeft(x, Vector128.CreateScalar((long)_amount).As<long, T>())`, and I just want to ensure the latter is getting hoisted correctly.

Reply:
With `ymm1` being initialized outside the loop, rather than directly before `vpslld`, correct?

Reply:
I actually don't see any loads into ymm1 at all; I do see a single `vmovd    xmm1, r9` that I assume is initializing half of ymm1?  But then, yes, lots of sequences like this:
```
       vmovups  ymm3, ymmword ptr [rcx]
       vpslld   ymm3, ymm3, ymm1
       vmovups  ymm4, ymmword ptr [rcx+0x20]
       vpslld   ymm4, ymm4, ymm1
       vmovups  ymm5, ymmword ptr [rcx+0x40]
       vpslld   ymm5, ymm5, ymm1
       vmovups  ymm6, ymmword ptr [rcx+0x60]
       vpslld   ymm6, ymm6, ymm1
```

Reply:
> I do see a single `vmovd    xmm1, r9`

Yep, this is it. Instructions prefixed with `v` that only operate on `xmm` implicitly clear the upper 128-bits of the `ymm` register implicitly (and upper 384-bits for `zmm`). Same goes for those operating on `ymm` clearing the upper 256-bits of `zmm`.

`vmovd` is doing `CreateScalar` and so the lower 64-bits are initialized to `r9` and upper 64-bits of `xmm` are implicitly zeroed.

> But then, yes, lots of sequences like this:

Looks as expected, thanks for confirming!